### PR TITLE
feat: assignment 4 circular dependencies

### PR DESCRIPTION
**Pull request by aristovinceselvanathan**

### CODE REVIEW

File: CSharp_Assignments.sln
Comment (sandeep-sol) [inline CSharp_Assignments.sln:1]: The assignment document says that the solution should be named as `CSharpConsoleSolution` but it's different here.

File: src/Assignment4/ProjectAGreetingsApp/ProgramA.cs [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+        /// <summary>
+        /// Main method prints the greeting message of Hello World and It creates the object of B and pass it method of class C.
+        /// Main method starts by the asking the two inputs from the user and checks for the input mismatch.
+        /// Main method asks for the operations that are to be performed. The loop runs as long as user terminate by the keyword.
+        /// </summary>
+        /// <param name="args">It is string array that returns from the command line interface</param>
+        public static void Main(string[] args)
+        {
+            int input1, input2, option;
+            string isNumber1, isNumber2;
```
Comment (sandeep-sol) [inline src/Assignment4/ProjectAGreetingsApp/ProgramA.cs]: The names should reflect what it does. Here it's not really making sense.

File: src/Assignment4/ProjectAGreetingsApp/ProgramA.cs [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+                }
+
+                Console.Write("Enter the Input2 : ");
+                isNumber2 = Console.ReadLine();
+                if (!int.TryParse(isNumber2, out input2))
+                {
+                    userInterface.InvalidNumberWarning("Input 2");
+                    continue;
+                }
+                else
```
Comment (sandeep-sol) [inline src/Assignment4/ProjectAGreetingsApp/ProgramA.cs]: Does having `else` here is really useful?

File: src/Assignment4/ProjectAGreetingsApp/ProgramA.cs [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+                {
+                    Console.Write("Enter the Option : 1.Addition 2.Subtraction 3.Multiplication 4.Division 5.Exit : ");
+                    if (!int.TryParse(Console.ReadLine(), out option) || (option > 6 || option < 1))
```
Comment (sandeep-sol) [inline src/Assignment4/ProjectAGreetingsApp/ProgramA.cs]: Nice one!!

File: src/Assignment4/ProjectBMathApp/ProgramB.cs [hunk 4]
```diff
@@ -0,0 +1,69 @@
+﻿namespace ProjectBMathApp
+{
+    using ProjectEHelperApp;
+
+    /// <summary>
+    /// ProgramB Class that implements the InterfaceE
+    /// </summary>
+    public class ProgramB : InterfaceE
```
Comment (sandeep-sol) [inline src/Assignment4/ProjectBMathApp/ProgramB.cs:8]: What do you expect this class to do?

If it just exposes functions for doing Math operations, then what is the need for main method?

File: src/Assignment4/ProjectCDisplayApp/ProgramC.cs [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+        public static void Main(string[] args)
+        {
+            Console.WriteLine("Hello, This is Display App Class");
+        }
+
+        /// <summary>
+        /// Method is help in get object of the ProgramB from ProgramA. To be used in the ProgramC.
+        /// </summary>
+        /// <param name="t1">It takes the reference of the InterfaceE</param>
+        public void Get(InterfaceE t1)
```
Comment (sandeep-sol) [inline src/Assignment4/ProjectCDisplayApp/ProgramC.cs]: Please update this function. This is not understandable.

File: src/Assignment4/ProjectCDisplayApp/ProgramC.cs [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+        {
+            temp = t1;
+        }
+
+        /// <summary>
+        /// Method display that display the result based on the operations that user choose.
```
Comment (sandeep-sol) [inline src/Assignment4/ProjectCDisplayApp/ProgramC.cs]: Typo mistake in the summary.

File: src/Assignment4/ProjectCDisplayApp/ProgramC.cs [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+        /// </summary>
+        /// <param name="input1">It takes the input1</param>
```
Comment (sandeep-sol) [inline src/Assignment4/ProjectCDisplayApp/ProgramC.cs]: This is a description of what the parameter is and not about what argument this function takes. Those 2 have completely different meanings.

File: src/Assignment4/ProjectCDisplayApp/ProgramC.cs [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+    using ProjectDUtilityApp;
+    using ProjectEHelperApp;
+
+    /// <summary>
+    /// ProgramC Class  used as helper class to access the ProgramB methods in ProgramC.
+    /// It keeps the object of the ProgramB Class in the reference of the interface in InterfaceE.
+    /// </summary>
+    public class ProgramC
+    {
+        private static InterfaceE temp;
```
Comment (sandeep-sol) [inline src/Assignment4/ProjectCDisplayApp/ProgramC.cs]: What is `temp`?

File: src/Assignment4/ProjectEHelperApp/InterfaceE.cs [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+    /// <summary>
+    /// InterfaceE Class
+    /// </summary>
+    public interface InterfaceE
+    {
+        /// <summary>
+        /// Main method that prints the Hello World
+        /// </summary>
+        /// <param name="args">It is string array that returns from the command line interface</param>
+        public static void Main(string[] args)
```
Comment (sandeep-sol) [inline src/Assignment4/ProjectEHelperApp/InterfaceE.cs]: Why is there a `Main` method in an interface?

**Review by sandeep-sol — CHANGES_REQUESTED** [review]

File: CSharp_Assignments.sln
Comment (aristovinceselvanathan) [reply CSharp_Assignments.sln:1]: I have resolved it.

File: src/Assignment4/ProjectAGreetingsApp/ProgramA.cs (hunk 1 above)
Comment (aristovinceselvanathan) [reply src/Assignment4/ProjectAGreetingsApp/ProgramA.cs]: I Have changed the variable name to be more meaningful.

File: src/Assignment4/ProjectBMathApp/ProgramB.cs (hunk 4 above)
Comment (aristovinceselvanathan) [reply src/Assignment4/ProjectBMathApp/ProgramB.cs:8]: I have removed the main method

File: src/Assignment4/ProjectCDisplayApp/ProgramC.cs (hunk 5 above)
Comment (aristovinceselvanathan) [reply src/Assignment4/ProjectCDisplayApp/ProgramC.cs]: I have named the method to be meaningful.

File: src/Assignment4/ProjectCDisplayApp/ProgramC.cs (hunk 7 above)
Comment (aristovinceselvanathan) [reply src/Assignment4/ProjectCDisplayApp/ProgramC.cs]: Parameter comments are described in detail.

File: src/Assignment4/ProjectCDisplayApp/ProgramC.cs (hunk 8 above)
Comment (aristovinceselvanathan) [reply src/Assignment4/ProjectCDisplayApp/ProgramC.cs]: I have changed the name.

File: src/Assignment4/ProjectEHelperApp/InterfaceE.cs (hunk 9 above)
Comment (aristovinceselvanathan) [reply src/Assignment4/ProjectEHelperApp/InterfaceE.cs]: I have removed the main method.

File: src/Assignment4/ProjectAGreetingsApp/ProgramA.cs (hunk 2 above)
Comment (aristovinceselvanathan) [reply src/Assignment4/ProjectAGreetingsApp/ProgramA.cs]: I have changed the logic of the else block

File: src/Assignment4/ProjectCDisplayApp/ProgramC.cs (hunk 6 above)
Comment (aristovinceselvanathan) [reply src/Assignment4/ProjectCDisplayApp/ProgramC.cs]: Typo mistake is rectified.

**Review by sandeep-sol — APPROVED** [review]

**Pull request merged**